### ISSUE TITLE
Collect the rules of API design policy

### ISSUE THREAD

**Issue by mrkn** [mrkn/mxnet.rb#38]
We import Gluon API mainly from Python's implementation.
Make our Gluon API appropriate to Ruby's programming style, we need to collect the rules of API design.

## Collected rules

- [ ] Do not import function-style constructor from Python

**Comment by toddsundsted** [comment]
@mrkn what versions of Ruby do you want to target? are there any language changes in latter versions of Ruby that you want to avoid using?

**Comment by mrkn** [comment]
@toddsundsted Now I want to support 2.4 or higher because 2.3 is going to be EOL soon.
Moreover, I think it is enough we care the living versions of rubies only until we can release the stable version of mxnet.rb.

**Comment by tjad** [comment]
Activations - No class implementations, however it is possible to pass in name

- [ ] LeakyReLU
- [ ] PReLU
- [ ] ELU
- [ ] SELU
- [ ] GELU
- [ ] Swish


Layers

- [x] Sequential
- [x] HybridSequential
- [x] Dense
- [ ] Dropout
- [ ] BatchNorm
- [ ] Embedding
- [x] Flatten
- [ ] InstanceNorm
- [ ] LayerNorm
- [ ] GroupNorm
- [ ] Lambda
- [ ] HybridLambda

- [x] _Conv
- [ ] Conv1D
- [x] Conv2D
- [ ] Conv3D
- [ ] Conv1DTranspose
- [ ] Conv2DTranspose
- [ ] Conv3DTranspose
- [x] _Pooling
- [ ] MaxPool1D
- [x] MaxPool2D
- [ ] MaxPool3D
- [ ] AvgPool1D
- [ ] AvgPool2D
- [ ] AvgPool3D
- [ ] GlobalMaxPool1D
- [ ] GlobalMaxPool2D

- [ ] GlobalMaxPool3D
- [ ] GlobalAvgPool1D
- [ ] GlobalAvgPool2D
- [ ] GlobalAvgPool3D
- [ ] ReflectionPad2D

RNN
- [ ] GRU Cell
- [ ] LSTM Cell
- [ ] GRU Layer
- [ ] LSTM Layer

Loss

- [x] L2Loss
- [x] L1Loss
- [ ] SigmoidBinaryCrossEntropyLoss
- [x] SoftmaxCrossEntropyLoss
- [ ] KLDivLoss
- [ ] CTCLoss
- [ ] HuberLoss
- [ ] HingeLoss
- [ ] SquaredHingeLoss
- [ ] LogisticLoss
- [ ] TripletLoss
- [ ] PoissonNLLLoss
- [ ] CosineEmbeddingLoss
- [ ] SDMLLoss
